### PR TITLE
Added database(PostgreSQL/MySQL)/ cache(Redis) server versions

### PR DESCRIPTION
**Pull request by rajat641**
Before submitting pull request, please ensure that:

- [x] Every commit has a message which describes it
- [x] Every commit is needed on its own, if you have just minor fixes to previous commits, you can squash them
- [ ] Any code changes come with test case
- [ ] Any new functionality is covered by user documentation

### CODE REVIEW

File: weblate/utils/requirements.py [hunk 1]
```diff
@@ -169,10 +167,75 @@ def get_versions():
...
+            )
+
+    return result
+
+
+def get_cache_version():
+    import redis
+
+    try:
+        result = ("Redis", "https://redis.io/", str(redis.__version__))
```
Comment (nijel) [inline weblate/utils/requirements.py:222]: Isn't this version of the redis module?

File: weblate/utils/requirements.py [hunk 2]
```diff
@@ -83,9 +84,6 @@
     "boto3",
     "zeep",
     "aeidon",
-    "iniparse",
-    "mysqlclient",
-    "pyparsing",
```
Comment (nijel) [inline weblate/utils/requirements.py:88]: The removal looks wrong.

File: weblate/utils/requirements.py [hunk 3]
```diff
@@ -169,10 +167,75 @@ def get_versions():
     return result
 
 
+def get_version_number_db(database, version):
+    """
+        Returns version of database in format as (Major Version.Minor Version.Revision)
+        from an integer. To know more about database version , refer <https://www.psycopg.org/docs/connection.html>
+    """
+    if database is "POSTGRESQL":
```
Comment (nijel) [inline weblate/utils/requirements.py:175]: This code seems used for postgres only, maybe it doesn't need the database arg and condition?

File: weblate/utils/requirements.py [hunk 4]
```diff
@@ -169,10 +167,75 @@ def get_versions():
...
+        major_version = version[0:2]
```
Comment (nijel) [inline weblate/utils/requirements.py:176]: This will break on postgres 9.x, the version there is for example "90209".

File: weblate/utils/requirements.py [hunk 5]
```diff
@@ -169,10 +170,81 @@ def get_versions():
...
+
+def get_cache_version():
+    try:
+        import os
+
+        cmd = "redis-server --version"
+        result = (
+            "Redis",
+            "https://redis.io/",
+            str(os.popen(cmd).read().split(" ")[2][2:]),
```
Comment (nijel) [inline weblate/utils/requirements.py:230]: There is no reason to have redis installed and running on same machine. Actually, in all Docker deployments it is not installed.

What should work instead (after checking that cache backend is redis) is: `cache.client.get_client().info()['redis_version']`

Comment (rajat641) [reply weblate/utils/requirements.py:230]: Hi @nijel , 
I am checking cache backend as Redis through `settings.CACHES.get('default').get('BACKEND') == "django_redis.cache.RedisCache"`.
Also, I'm not sure what to import  for `cache.client.get_client().info()['redis_version']` to work??
I have tried checking with `import django_redis.cache` but it's throwing some error or may I'm importing something wrong.

File: weblate/utils/requirements.py (hunk 4 above)
Comment (rajat641) [reply weblate/utils/requirements.py:180]: I have fixed it.

File: weblate/utils/requirements.py (hunk 3 above)
Comment (rajat641) [reply weblate/utils/requirements.py]: I have fixed it as well.

File: weblate/utils/requirements.py [hunk 6]
```diff
@@ -169,10 +171,88 @@ def get_versions():
...
+            )
+
+    return result
+
+
+def get_cache_version():
+    try:
+        if (
+            settings.CACHES.get("default").get("BACKEND")
+            == "django_redis.cache.RedisCache"
```
Comment (rajat641) [inline weblate/utils/requirements.py:234]: I am checking the cache backend here and if backend is not redis. I am raising an exception. Please let me know it's correct.

File: weblate/utils/requirements.py (hunk 2 above)
Comment (rajat641) [reply weblate/utils/requirements.py]: I used pycharm code refactor tool. It modified it somehow. It's fixed now.

File: weblate/utils/requirements.py (hunk 3 above)
Comment (rajat641) [reply weblate/utils/requirements.py]: I'll use the flake8 and `pre-commit run --all` from now on .

File: weblate/utils/requirements.py (hunk 5 above)
Comment (nijel) [reply weblate/utils/requirements.py]: `from django.core.cache import cache` should do it.

File: weblate/utils/requirements.py [hunk 7]
```diff
@@ -169,10 +171,88 @@ def get_versions():
     return result
 
 
+def get_version_number_db(version):
+    """Returns version of database in format as (Major Version.Minor Version.Revision).
+
+    from an integer. To know more about database version,
+    refer <https://www.psycopg.org/docs/connection.html>.
+    """
```
Comment (nijel) [inline weblate/utils/requirements.py:179]: Can you please add tests for this? The code looks to complex for what it does and it's hard to review that it works properly. Having tests would make the verification it works correctly. It should verify all corner case you are handling (`90209` -> 9.2.9, `120003` -> 12.3, `120000` -> 12).

Maybe it would be better to avoid this postporcessing and execute `SHOW server_version;` SQL directly?

File: weblate/utils/requirements.py (hunk 5 above)
Comment (rajat641) [reply weblate/utils/requirements.py]: Yes, it works. Thanks

File: weblate/utils/requirements.py [hunk 8]
```diff
@@ -169,10 +171,88 @@ def get_versions():
...
+                "Failed to get a database version. please install a database."
+            )
+
+    return result
+
+
+def get_cache_version():
+    try:
+        if (
+            settings.CACHES.get("default").get("BACKEND")
```
Comment (nijel) [inline weblate/utils/requirements.py:233]: Chaining .get is probably not a good idea here - in case "default" is not present it will return None and next get will crash. So either rely on "default" being present or provide default value to get.

File: weblate/utils/requirements.py (hunk 7 above)
Comment (rajat641) [reply weblate/utils/requirements.py:179]: I'll try to avoid this post-processing and string manipulations. Will add try to execute `SHOW server_version` and parse its results.

File: weblate/utils/requirements.py (hunk 8 above)
Comment (rajat641) [reply weblate/utils/requirements.py:233]: Will fix this.

Comment (rajat641) [reply weblate/utils/requirements.py:233]: `settings.CACHES.get("default", {}).get("BACKEND")` would work fine I guess.

File: weblate/utils/requirements.py (hunk 7 above)
Comment (rajat641) [reply weblate/utils/requirements.py:179]: @nijel , I think as per documentation 120000->12.0 
image

File: weblate/utils/requirements.py (hunk 5 above)
Comment (rajat641) [reply weblate/utils/requirements.py]: @nijel - the code `from django.core.cache import cache` looks magical. Can you refer a source where its available so that I can refer it for the future?

Comment (nijel) [reply weblate/utils/requirements.py]: https://docs.djangoproject.com/en/3.0/topics/cache/#the-low-level-cache-api

File: weblate/utils/requirements.py (hunk 7 above)
Comment (nijel) [reply weblate/utils/requirements.py]: You're right with 12.0, these were mostly given as examples without deep investigation. Still, I think it's better to avoid such code if possible :-).

File: weblate/utils/requirements.py (hunk 1 above)
Comment (rajat641) [reply weblate/utils/requirements.py]: @nijel - This was fixed.

File: weblate/utils/requirements.py (hunk 5 above)
Comment (rajat641) [reply weblate/utils/requirements.py]: @nijel - Thanks for sharing. I read it over the weekend here.

File: weblate/utils/requirements.py (hunk 7 above)
Comment (rajat641) [reply weblate/utils/requirements.py]: @nijel - I have avoided the complex string manipulation that I did before. Now, I am using `connection.cursor().connection.get_server_info().split("-", 1)[0]` to parse MySQL version by quering the database and a similar approach in PostgreSQL

Comment (rajat641) [reply weblate/utils/requirements.py]: Please let me if there is any better way to do things. The previous one was not definitely hackish.

Comment (rajat641) [reply weblate/utils/requirements.py]: Also, after the code changes `weblate list_versions` command works fine. I have tested it after databases.

File: weblate/utils/requirements.py [hunk 9]
```diff
@@ -169,10 +172,66 @@ def get_versions():
     return result
 
 
+def get_db_version():
+    if connection.vendor.upper() == "POSTGRESQL":
+        try:
+            cursor = connection.cursor()
+            cursor.execute("SHOW server_version")
+            version = cursor.fetchall()[0]
```
Comment (nijel) [inline weblate/utils/requirements.py:180]: This should be better:

```suggestion
            version = cursor.fetchone()
```

File: weblate/utils/requirements.py [hunk 10]
```diff
@@ -169,10 +172,66 @@ def get_versions():
...
+def get_cache_version():
+    try:
+        if (
+            settings.CACHES.get("default", {}).get("BACKEND")
+            == "django_redis.cache.RedisCache"
+        ):
+            version = cache.client.get_client().info()["redis_version"]
+            result = ("Redis", "https://redis.io/", version)
+        else:
+            raise ImproperlyConfigured("please install a redis server.")
```
Comment (nijel) [inline weblate/utils/requirements.py:218]: I'm not sure about this exception. Weblate will probably work just fine with other cache backend, so raising an exception seems too strict here.

Comment (rajat641) [reply weblate/utils/requirements.py:218]: I will remove this exception. But I'm not sure, how to get the cache version/name for some other backend.There can be many such services. OR  I have to show only cache version if the backend is `django_redis.cache.RedisCache`

File: weblate/utils/requirements.py (hunk 9 above)
Comment (rajat641) [reply weblate/utils/requirements.py:180]: I'll fix this and update it.

File: weblate/utils/requirements.py (hunk 10 above)
Comment (nijel) [reply weblate/utils/requirements.py:218]: There is no reason to cover all options, getting it only for redis is good enough as that will cover most use cases. But it should not break in others.

File: weblate/utils/requirements.py (hunk 9 above)
Comment (rajat641) [reply weblate/utils/requirements.py]: @nijel - This is fixed.

File: weblate/utils/requirements.py (hunk 10 above)
Comment (rajat641) [reply weblate/utils/requirements.py]: @nijel - I have removed the `ImproperlyConfigured` exception. Also, I tested the code with another cache namely Memcached and its working fine.

**Comment by nijel** [comment]
Merged, thanks for your contribution! I've made some further improvements to the code in 5ce7ecb634

**Comment by rajat641** [comment]
Thanks, @nijel . I have seen the code changes. It was a great learning experience.
